In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("/content/WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [3]:
df.head(5)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [4]:
df.TotalCharges = pd.to_numeric(df.TotalCharges, errors='coerce')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [6]:
df.dropna(how='any', inplace=True)

In [7]:
X = df.drop(['customerID', 'Churn'], axis=1)
y = df.Churn.values

In [8]:
X.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'MonthlyCharges', 'TotalCharges'],
      dtype='object')

In [9]:
X = pd.get_dummies(X, columns=['gender', 'Partner', 'Dependents',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod'], drop_first=True)

In [14]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [15]:
len(X_train), len(X_test)

(5274, 1758)

In [16]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_sc = scaler.fit_transform(X_train)
X_test_sc = scaler.transform(X_test)

***kNN Classifier***

In [17]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()
knn.fit(X_train_sc, y_train)

KNeighborsClassifier()

In [18]:
y_pred_kn = knn.predict(X_test_sc)

In [20]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred_kn)*100)

75.36973833902162


***kNN***
**Accuracy Score**
->75.36%

---

***Decision Tree Classifier***


In [21]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

DecisionTreeClassifier()

In [22]:
y_pred_dt = dt.predict(X_test)

In [23]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred_dt)*100)

71.4448236632537


***Decision Tree Classifier -> Accuracy Score-> 71.44%***

---



***Random Forest Classifier***

In [27]:
from sklearn.ensemble import RandomForestClassifier
model_rf = RandomForestClassifier(n_estimators=200)
model_rf.fit(X_train, y_train)

RandomForestClassifier(n_estimators=200)

In [28]:
y_pred_rf = model_rf.predict(X_test)

In [29]:
from sklearn.metrics  import accuracy_score
print(accuracy_score(y_test, y_pred_rf)*100)

78.44141069397043


***Random Forest Classifier -> Accuracy Score-> 78.44%***

---



***Naive Bayes***

In [30]:
from sklearn.naive_bayes import BernoulliNB
model_nb = BernoulliNB()
model_nb.fit(X_train, y_train)

BernoulliNB()

In [31]:
from sklearn.metrics import accuracy_score
y_pred_nb = model_nb.predict(X_test)
print(accuracy_score(y_test, y_pred_nb)*100)

70.81911262798634


***Naive Bayes Classifier -> Accuracy Score-> 70.8%***

---



***SVM Classifier***

In [34]:
from sklearn.svm import SVC
model_svm = SVC()
model_svm.fit(X_train_sc, y_train)

SVC()

In [35]:
from sklearn.metrics import accuracy_score
y_pred_svm = model_svm.predict(X_test_sc)
print(accuracy_score(y_test, y_pred_svm)*100)

79.29465301478953


***SVM Classifier -> Accuracy Score-> 79.29%***

---



***Logistic Regression Classifier***

In [37]:
from sklearn.linear_model import LogisticRegression
model_lr = LogisticRegression()
model_lr.fit(X_train_sc, y_train)

LogisticRegression()

In [38]:
from sklearn.metrics import accuracy_score
y_pred_lr = model_lr.predict(X_test_sc)
print(accuracy_score(y_test, y_pred_lr)*100)

79.01023890784982


***Logistic Regression Classifier -> Accuracy Score-> 79.01%***

---

